modular_skyrat/modules/colony_fabricator/code/machines/arc_furnace.dm

```diff
@@ -0,0 +1,194 @@
+	layer = BELOW_OBJ_LAYER
+	density = TRUE
+	circuit = null
+	light_color = LIGHT_COLOR_BRIGHT_YELLOW
+	light_power = 10
+	flags_1 = NODECONSTRUCT_1
+	active_power_usage = BASE_MACHINE_ACTIVE_CONSUMPTION * 10 // This baby consumes so much power
+	/// The item we turn into when repacked
+	var/repacked_type = /obj/item/flatpacked_machine/arc_furnace
+	/// If the furnace is currently working on smelting something
+	var/operating = FALSE
+	/// Image for the radial eject button
+	var/static/radial_eject = image(icon = 'icons/hud/radial.dmi', icon_state = "radial_eject")
+	/// Image for the radial use button
+	var/static/radial_use = image(icon = 'icons/hud/radial.dmi', icon_state = "radial_use")
+	/// Radial options for using the arc furnace
+	var/static/list/radial_options = list("eject" = radial_eject, "use" = radial_use)
+	/// Soundloop for while we are smelting ores
+	var/datum/looping_sound/arc_furnace_running/soundloop
+
+/obj/machinery/arc_furnace/Initialize(mapload)
+	. = ..()
+	soundloop = new(src, FALSE)
+	AddElement(/datum/element/repackable, repacked_type, 2 SECONDS)
+	if(!mapload)
+		flick("arc_furnace_deploy", src)
+
+/obj/machinery/arc_furnace/examine(mob/user)
+	. = ..()
+	if(length(contents))
+		. += span_notice("It has <b>[contents[1]]</b> sitting in it.")
+
+/obj/machinery/arc_furnace/deconstruct(disassembled = TRUE)
+	eject_contents()
+	return ..()
+
+/obj/machinery/arc_furnace/update_appearance()
+	. = ..()
+	cut_overlays()
+
+	if(length(contents))
+		var/image/overlayed_item = image(icon = contents[1].icon, icon_state = contents[1].icon_state)
+		overlayed_item.transform = matrix(, 0, 0, 0, 0.8, 0)
+		add_overlay(overlayed_item)
+
+	var/image/furnace_front_overlay = image(icon = icon, icon_state = "[operating ? "[base_icon_state]_overlay_active" : "[base_icon_state]_overlay"]")
+	add_overlay(furnace_front_overlay)
+
+/obj/machinery/arc_furnace/attackby(obj/item/attacking_item, mob/living/user, params)
+	if(operating)
+		balloon_alert(user, "furnace busy")
+		return TRUE
+
+	if(length(contents))
+		balloon_alert(user, "furnace full")
+		return TRUE
+
+	if(istype(attacking_item, /obj/item/stack/ore))
+		attacking_item.forceMove(src)
+		balloon_alert(user, "ore added")
+		update_appearance()
+		return TRUE
+
+	return ..()
+
+/obj/machinery/arc_furnace/ui_interact(mob/user)
+	. = ..()
+
+	if(operating || !user.can_perform_action(src, ALLOW_SILICON_REACH))
+		return
+	if(isAI(user) && (machine_stat & NOPOWER))
+		return
+
+	if(!length(contents))
+		balloon_alert(user, "it's empty!")
+		return
+
+	var/choice = show_radial_menu(user, src, radial_options, require_near = !issilicon(user))
+
+	// post choice verification
+	if(operating || !user.can_perform_action(src, ALLOW_SILICON_REACH))
+		return
+	if(isAI(user) && (machine_stat & NOPOWER))
+		return
+
+	usr.set_machine(src) // What does this even do??
+	switch(choice)
+		if("eject")
+			eject_contents()
+		if("use")
+			smelt_it_up(user)
```

Comment:
The two options could be defines, so they're easier to maintain. Otherwise it's just magic strings.